redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+/// including retrieving node information, refreshing slot mapping, and routing commands to specific nodes.
+///
+/// The `ScanState` struct represents the state of a scan operation in a Redis cluster.
+/// It holds information about the current scan state, including the cursor position, scanned slots map,
+/// slots in the scan map, node being scanned, and node's epoch.
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use crate::aio::ConnectionLike;
+use crate::cluster_async::{ClusterConnInner, Connect, InnerCore};
+use crate::cluster_routing::SlotAddr;
+use crate::{cmd, from_redis_value, Cmd, ErrorKind, RedisError, RedisResult, Value};
+
+const BITS_PER_U64: usize = 64;
+const NUM_OF_SLOTS: usize = 16384;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ScanState {
+    pub cursor: u64,
+    pub scanned_slots_map: SlotsBitsArray,
+    pub slots_in_scan_map: SlotsBitsArray,
+    pub node_in_scan: String,
+    pub node_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    scan_state_ref: ScanStateWrapper,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<String>,
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_ref: ScanStateWrapper,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<String>,
+    ) -> Self {
+        Self {
+            scan_state_ref,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateWrapper {
+    scan_atate_ref: Arc<Option<ScanState>>,
+}
+
+impl PartialEq for ScanStateWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.is_none() == other.is_none()
+    }
+}
+
+impl Default for ScanStateWrapper {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ScanStateWrapper {
+    fn from(scan_state: &ScanState) -> Self {
+        Self {
+            scan_atate_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateWrapper`.
+    ///
+    /// This method initializes the `ScanStateWrapper` with a read-write lock
+    /// that is initially set to `None`.
+    pub fn new() -> Self {
+        Self {
+            scan_atate_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
+    pub fn is_none(&self) -> bool {
+        self.scan_atate_ref.is_none()
+    }
+
+    /// Returns a clone of the scan state, if it exists.
+    pub fn get_state_from_wrrraper(&self) -> ScanState {
+        self.scan_atate_ref.as_ref().clone().unwrap()
```

Comment:
using .unwrap() is a bad practice - it can make the user app to panic.
Instead of this function, derive/implement the clone trait for ScanStateWrapper.

Reply:
Also - when is the user is going to clone the scan state? and why?

Reply:
I see that you use this function internally to get the inner state. This function is `pub` which implies it is a user facing API

Reply:
change to pub(crate)

Reply:
Its not public, it is for internal use. i need in some places in the code the stateref itself and not a ref to the ref. 
For the first part - i will.

Reply:
when you use 'pub' it implies it can and should be used by users. use pub(crate) instead, or lets think of different options to get the state